internal/sbi/httpcallback/api_handle_dereg_notification.go

```diff
@@ -0,0 +1,105 @@
+	"github.com/free5gc/amf/internal/logger"
+	"github.com/free5gc/amf/internal/sbi/consumer"
+	"github.com/free5gc/openapi"
+	"github.com/free5gc/openapi/models"
+)
+
+func HTTPAmfHandleDeregistrationNotification(c *gin.Context) {
+	// TS 23.502 - 4.2.2.2.2 - step 14d
+	logger.CallbackLog.Infoln("Handle Deregistration Notification")
+
+	var deregData models.DeregistrationData
+
+	requestBody, err := c.GetRawData()
+	if err != nil {
+		logger.CallbackLog.Errorf("Get Request Body error: %+v", err)
+		problemDetails := models.ProblemDetails{
+			Title:  "System failure",
+			Status: http.StatusInternalServerError,
+			Detail: err.Error(),
+			Cause:  "SYSTEM_FAILURE",
+		}
+		c.JSON(http.StatusInternalServerError, problemDetails)
+		return
+	}
+
+	err = openapi.Deserialize(&deregData, requestBody, "application/json")
+	if err != nil {
+		problemDetails := "[Request Body] " + err.Error()
+		rsp := models.ProblemDetails{
+			Title:  "Malformed request syntax",
+			Status: http.StatusBadRequest,
+			Detail: problemDetails,
+		}
+		logger.CallbackLog.Errorln(problemDetails)
+		c.JSON(http.StatusBadRequest, rsp)
+		return
+	}
+
+	ueid := c.Param("ueid")
+	amfSelf := amf_context.GetSelf()
+	ue, ok := amfSelf.AmfUeFindByUeContextID(ueid)
+	if !ok {
+		logger.CallbackLog.Errorf("AmfUe Context[%s] not found", ueid)
+		problemDetails := &models.ProblemDetails{
+			Status: http.StatusNotFound,
+			Cause:  "CONTEXT_NOT_FOUND",
+		}
+		c.JSON(http.StatusNotFound, problemDetails)
+		return
+	}
+
+	switch deregData.DeregReason {
+	case models.DeregistrationReason_UE_INITIAL_REGISTRATION:
+		// TS 23.502 - 4.2.2.2.2 General Registration
+		// Invokes the Nsmf_PDUSession_ReleaseSMContext for the corresponding access type
+		ue.SmContextList.Range(func(key, value interface{}) bool {
+			smContext := value.(*amf_context.SmContext)
+
+			if smContext.AccessType() == deregData.AccessType {
+				var problemDetails *models.ProblemDetails
+				problemDetails, err = consumer.SendReleaseSmContextRequest(ue, smContext, nil, "", nil)
+				if problemDetails != nil {
+					ue.GmmLog.Errorf("Release SmContext Failed Problem[%+v]", problemDetails)
+				} else if err != nil {
+					ue.GmmLog.Errorf("Release SmContext Error[%v]", err.Error())
+				}
+			}
+
+			return true
+		})
+	case models.DeregistrationReason_SUBSCRIPTION_WITHDRAWN:
+		// TS 23.502 - 4.2.2.3.3 Network-initiated Deregistration
+		// The AMF executes Deregistration procedure over the access(es) the Access Type indicates
+
+		// Use old AMF as the backup AMF
+		backupAmfInfo := models.BackupAmfInfo{
+			BackupAmf: amfSelf.Name,
+			GuamiList: amfSelf.ServedGuamiList,
+		}
+		ue.UpdateBackupAmfInfo(backupAmfInfo)
+	}
+
+	// TS 23.502 - 4.2.2.2.2 General Registration
+	// The old AMF should clean the UE context
+	// The AMF also unsubscribes with the UDM using Nudm_SDM_Unsubscribe service operation.
+	err = gmm_common.PurgeSubscriberData(ue, deregData.AccessType)
```

Comment:
For SDMUnsubscribe in PurgeSubscriberData, 23.502 4.2.2.2.2 step 14.e mentioned that
`If old AMF does not have UE context for another access type (i.e. non-3GPP access), the Old AMF unsubscribes with the UDM for subscription data using Nudm_SDM_unsubscribe`
Should this be checked here?

Reply:
I followed the TS 23.502 - 4.2.2.3.3 Network-initiated Deregistration, 
> [Conditional] If the Deregistration procedure is triggered by UDM (Step 1), the AMF acknowledges the Nudm_UECM_DeRegistrationNotification to the UDM. The AMF also unsubscribes with the UDM using Nudm_SDM_Unsubscribe service operation.

However, it seems that we should also consider whether old AMF have UE context for non-3GPP access.
I am not sure which elements in the AmfUe context belong to non-3GPP, maybe you can enlighten me. :)

Reply:
Generally, contexts in AmfUe without `map[models.AccessType]Xxx` are contexts shared by both 3gpp and non-3gpp. In other words, contexts with  `map[models.AccessType]Xxx` may indicate the existance of non-3gpp contexts. 
I think one way to judge whether amf has non-3gpp access may be checking the existence of ameUe.State[N3GPP], but that requires refactoring for amfUe.init() since it initializes ameUe.State[N3GPP] with Deregistered state instead of nil.😂

I think this can be added as TODO comment if that may take time to achieve currently, thanks.